maro/simulator/scenarios/finance/business_engine.py

```diff
@@ -0,0 +1,653 @@
+                    executing = False
+                else:
+                    rets.append(ret)
+            else:
+                executing = False
+                ret = None
+
+        return rets
+
+    def trade(self, action: Action):
+        rets = []
+        market_price = self._stocks[action.item].opening_price
+        market_volume = self._stocks[action.item].trade_volume
+        remaining_money = self._account.remaining_money
+
+        slippage = ByMoneySlippage(self._conf["trade_constraint"]["slippage"])
+        commission = ByMoneyCommission(self._conf["trade_constraint"]["commission"], 5)
+        tax = StampTaxCommission(self._conf["trade_constraint"]["close_tax"])
+
+        executing_volume = action.amount
+        max_trade_volume = self._conf["trade_constraint"]["max_trade_percent"] * market_volume
+        if executing_volume > max_trade_volume:
+            executing_volume = max_trade_volume
+        odd_shares = self._stocks[action.item].account_hold_num % self._conf["trade_constraint"]["min_sell_unit"]
+        # print("action.amount", action.amount)
+
+        if action.direction == OrderDirection.sell:
+            executing_volume = min(executing_volume, self._stocks[action.item].account_hold_num)
+
+        adjusted = True
+
+        while adjusted:
+            adjusted = False
+
+            executing_price = market_price
+
+            executing_price = slippage.execute(action.direction, executing_volume, executing_price, market_volume)
+
+            executing_commission = commission.execute(action.direction, executing_price, executing_volume)
+
+            executing_tax = tax.execute(action.direction, executing_price, executing_volume)
+
+            # constraint
+            if not self.validate_constrant(odd_shares, action.direction, executing_price, executing_volume):
+                executing_volume = self.adjust_constraint(
+                    odd_shares, action.direction, executing_price, executing_volume
+                )
+                adjusted = True
+                # print(f"Adjust volume for constraint:{executing_volume}")
+                continue
+
+            # money
+            if not self.validate_trade(
+                action.direction, executing_price, executing_volume,
+                executing_commission, executing_tax, remaining_money
+            ):
+                executing_volume = self.adjust_trade(
+                    action.direction, executing_price, executing_volume,
+                    executing_commission, executing_tax, remaining_money
+                )
+                adjusted = True
+                # print(f"Adjust volume for trade:{executing_volume}")
+                continue
+
+        if executing_volume > 0:
+            ret = TradeResult(executing_volume, executing_price, executing_commission, executing_tax)
+            rets.append(ret)
+        else:
+            pass
+
+        return rets
+
+    def validate_constrant(
```

Comment:
rename "constrant" to "constraint"

Reply:
rename ``validate_constraint`` to ``validate_trade_on_volume_unit``?
rename ``adjust_constraint`` to ``adjust_trade_on_volume_unit``?

Any other proposal?

Reply:
good idea